enhancements/scheduling/load-aware-vm-rebalancing-through-descheduler.md

```diff
@@ -0,0 +1,472 @@
+
+### Update LowNodeUtilization plugin to consume Prometheus metrics
+
+Currently, the descheduler's `LowNodeUtilization` plugin classifies node utilization based on Kubernetes metrics. However, these metrics primarily capture CPU and memory usage at the node and pod levels, which are not well-suited for VMs. By consuming Prometheus metrics, the plugin can leverage a broader range of data that better aligns with VM-specific needs.
+
+The existing plugin code is already structured to support additional resource clients, making integration straightforward. To simplify implementation, the plugin will accept a single metric entry per node, with values in the <0;1> interval—where higher value indicates higher node utilization. Since resource thresholds are expressed as percentages by default, this classification remains intuitive and easy to interpret.
+
+In the current iteration of the design, PSI pressure metrics were selected. Due to their nature, it is not possible to predict how evicting a single VM will impact a node's reported PSI pressure. Therefore, the plugin will be designed to evict only one VM per over-utilized node per cycle by default with option to evict more. Since it is safe to assume that each evicted VM will undergo live migration and the number of evictions per node is limited, running the descheduling cycle more frequently (e.g., every 3 minutes) will approximate the effect of evicting multiple VMs within a single cycle.
+
+### Soft-taint overutilized nodes to make them less attractive for scheduling
+
+The default Kubernetes scheduler in OpenShift cannot currently be extended with load-aware capabilities. However, applying soft-taints (taints with the `PreferNoSchedule` effect) to over-utilized nodes can guide the scheduler to avoid placing VM pods on these nodes, which would otherwise be selected as targets.
+
+A new taint controller applying the soft taints requires permission to update a node by adding or removing taints. Typically, node update RBAC allows modifications to any part of a node object. However, this permission must be restricted to allow manipulation only for specific taints, ensuring more granular control over node updates. The CEL `ValidatingAdmissionPolicy` can be used to enforce this restriction.
+
+At present, only a single soft-taint is expected to be applied. However, in certain scenarios, applying multiple soft-taints could enhance overall rebalancing by providing more granular scheduling guidance.
+
+The taint controller needs to be allowed to switch to a mode where it only removes the soft-taint(s) from all affected nodes. In case an administrator decides to disable the load-aware functionality.
+
+In general, eviction, scheduling, and tainting steps can occur at any time, making VM rebalancing inherently non-deterministic. As a result, the process may not always converge to an optimal state or may take a long time to do so.
```

Comment:
if i understand it correctly, the steps involved are:
- a1) descheduler plugin reads the PSI metrics, andclassifies a node (overcommitted, undercommitted, normal) 
- a2) descheduler plugin evicts a Pod (a VM in this case) based on results from `a`
- b) taint controller applies soft tainting label to an overcommitted node
- c) scheduler schedules the evicted Pod to a new node

a, b and c are asyn steps, is this a correct summary of the overall flow?